libraries/AdaptiveExpressions/ExpressionFunctions.cs

```diff
@@ -4265,13 +4528,54 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
                 new ExpressionEvaluator(ExpressionType.UriComponentToString, Apply(args => Uri.UnescapeDataString(args[0].ToString()), VerifyString), ReturnType.String, ValidateUnary),
 
                 // TODO: Is this really the best way?
-                new ExpressionEvaluator(ExpressionType.String, Apply(args => JsonConvert.SerializeObject(args[0]).TrimStart('"').TrimEnd('"')), ReturnType.String, ValidateUnary),
+                new ExpressionEvaluator(
+                    ExpressionType.String,
+                    (expression, state, options) =>
+                        {
+                            string result = null;
+                            var (args, error) = EvaluateChildren(expression, state, options);
+                            if (args.Count == 2 && !(args[1] is string))
+                            {
+                                error = $"the second argument {args[1]} should be a locale string.";
+                            }
+                            else
+                            {
+                                try
+                                {
+                                    var locale = args.Count == 2 ? new CultureInfo(args[1] as string) : DetermineLocaleInfo(state);
+                                    if (args[0].IsNumber())
+                                    {
+                                        result = Convert.ToDouble(args[0]).ToString(locale);
+                                    }
+                                    else if (args[0] is DateTime dt)
+                                    {
+                                        result = dt.ToString(DefaultDateTimeFormat, locale);
+                                    }
+                                    else if (args[0] is TimeSpan ts)
+                                    {
+                                        result = ts.ToString(DefaultTimeSpanFormat, locale);
+                                    }
+                                    else
+                                    {
+                                        result = JsonConvert.SerializeObject(args[0]).TrimStart('"').TrimEnd('"');
+                                    }
+                                }
+                                catch
+                                {
+                                    error = $"{args[1]} is not a valid locale for string";
+                                }
+                            }
+
+                            return (result, error);
+                        }, 
+                    ReturnType.String,
+                    expr => ValidateOrder(expr, new[] { ReturnType.String }, ReturnType.Object)),
```

Comment:
>expr => ValidateOrder(expr, new[] { ReturnType.String }, ReturnType.Object)) [](start = 20, length = 76)

Not sure I get the semantics here.  I should be able to do string('foo') or string('foo', 'en-us') or string(3000, 'fr-fr').  Where is the optional thing an object?

Reply:
for ValidateOrder method, the signature is expression, optional parameters array, fixed parameters. 
So new[] { ReturnType.String } (optional) comes before ReturnType.Object